[CDAP-13004] Adds PipelineDetailStore and PipelineDetail ActionCreator

tonybach opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-13004

This is the first step in rewriting the Pipeline Detail view top panel in React (https://issues.cask.co/browse/CDAP-12903). In this PR I have added:
- PipelineDetailStore, which is the Redux rewrite of `pipeline-detail-runs-store.js` and `pipeline-detail-nonruns-store.js` merged together
- PipelineDetails Action Creator, which is the Redux rewrite of `pipeline-detail-actions.js`

I have not changed references to the old Angular stores and action creator yet. That will come in the next PR.

tonybach committed:
[CDAP-13004] Adds PipelineDetailStore and PipelineDetail ActionCreator
- Adds PipelineDetails/store/ActionCreator as Redux rewrite of pipeline-detail-actions, and adds more API on React side
- Remove references to GLOBALS.etlBatch and GLOBALS.etlRealtime in PipelineDetailRunStore since we don't care about those anymoreg
- Merges pipelne-detail-nonruns-store to PipelineDetailRunStore, and some logic simplification
- Changes latest copyright years to 2018
- Changes POLL apis to REQUEST, and changes polling functions in Action Creator to just do a single request

elfenheart commented on cdap-ui/app/cdap/components/PipelineDetails/store/ActionCreator.js:
```diff
@@ -0,0 +1,257 @@
+  });
+};
+
+const pollRuns = (params) => {
```
I see these poll functions, but where are the polls being stopped?

tonybach commented:
@elfenheart Hmm I added that because we have a corresponding function in the Angular store (https://github.com/caskdata/cdap/blob/feature/rewrite-pipeline-detail-stores/cdap-ui/app/hydrator/services/detail/actions/pipeline-detail-actions.js#L37). 

This is used at https://github.com/caskdata/cdap/blob/feature/rewrite-pipeline-detail-stores/cdap-ui/app/hydrator/controllers/detail-ctrl.js#L68, where the function is just being called without the polls being stopped. This seems wrong, so I have removed the `pollRuns` function from my new store, and will address the issue in my next PR (https://github.com/caskdata/cdap/pull/9868).

elfenheart commented:
So we should not poll from the resource standpoint anymore. In angular, we tie a polling function to `$scope`, so when the `$scope` is destroyed, the poll will be stopped automatically. 

In react, you should use Rx.Observable.interval to create a poll inside the component, that way you can unsubscribe and that will stop the poll when the component will unmount. So things like getStatistic should not be polling from action creator

elfenheart commented on cdap-ui/app/cdap/components/PipelineDetails/store/index.js:
```diff
@@ -0,0 +1,368 @@
+
+      appConfig.pipelineName = app.name;
+      appConfig.pipelineType = app.artifact.name;
+      appConfig.pipelineDescription = app.description;
```
shouldn't need to duplicate these

elfenheart commented on cdap-ui/app/cdap/components/PipelineDetails/store/index.js:
```diff
@@ -0,0 +1,368 @@
+  pipelineName: '',
+  pipelineType: '',
+  pipelineDescription: '',
+  cloneConfig: {},
```
should rename this cloneConfig to just config

elfenheart commented on cdap-ui/app/cdap/components/PipelineDetails/store/index.js:
```diff
@@ -69,6 +69,8 @@ const DEFAULT_PIPELINE_DETAILS = {
   type: '',
   description: '',
+  artifactVersion: '',
+  artifactScope: '',
```
wouldn't this be better to just store as the artifact full object? combine type, artifactVersion, and artifactScope?

elfenheart commented on cdap-ui/app/cdap/components/PipelineDetails/store/index.js:
```diff
@@ -101,10 +103,13 @@ const pipelineDetails = (state = DEFAULT_PIPELINE_DETAILS, action = defaultActio
-        artifactScope: pipeline.artifact.scope,
         description: pipeline.description,
+        artifact: {
+          ...state.artifact,
```
this line is not needed

elfenheart reviewed (approved):
👍  LGTM
